src/viam/sdk/common/proto_t.hpp

```diff
@@ -0,0 +1,293 @@
+    // moved-from type
+    ProtoT(ProtoT&& other) noexcept
+        : vtable_(std::move(other.vtable_)), self_(std::move(other.self_), vtable_) {}
+
+    ProtoT(const ProtoT& other) : vtable_(other.vtable_), self_(other.self_, other.vtable_) {}
+
+    // Move assignment from other, leaving other in the "unspecified but valid state" of its
+    // moved-from type.
+    ProtoT& operator=(ProtoT&& other) noexcept {
+        ProtoT(std::move(other)).swap(*this);
+        return *this;
+    }
+
+    ProtoT& operator=(const ProtoT& other) {
+        ProtoT(other).swap(*this);
+        return *this;
+    }
+
+    ~ProtoT() {
+        self_.destruct(vtable_);
+    }
+
+    // Test equality of two types.
+    // Note that "intuitive" arithmetic equality is not supported, but could be.
+    // Thus, bool{false}, int{0}, and double{0.0} do not compare equal.
+    friend bool operator==(const ProtoT& lhs, const ProtoT& rhs) {
+        return lhs.vtable_.equal_to(lhs.self_.get(), rhs.self_.get(), rhs.vtable_);
+    }
+
+    void swap(ProtoT& other) {
+        self_.swap(vtable_, other.self_, other.vtable_);
+        std::swap(vtable_, other.vtable_);
+    }
+
+    // Construct from proto value
+    // This method is trivially templated to insulate Value from our API/ABI.
+    // In a translation unit which includes <google/protobuf/struct.pb.h> you can call
+    // this function on a Value instance without specifying a template parameter and it will "just
+    // work"
+    template <typename Value = google::protobuf::Value>
+    static ProtoT from_proto_value(const Value& v) {
+        return ProtoT(&v);
+    }
+
+    // Convert to protobuf Value.
+    friend void to_proto_value(const ProtoT& t, google::protobuf::Value* v);
+
+    // Obtain integer constant representing the stored data type.
+    int kind() const {
+        return vtable_.kind();
+    }
+
+    // Checks whether this ProtoT is an instance of type T.
+    template <typename T>
+    bool is_a() const;
+
+    // Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T* dyn_cast(ProtoT&);
+
+    // Checking cast to T, returns non-owning non-null pointer if argument is_a<T>()
+    template <typename T>
+    friend T const* dyn_cast(const ProtoT&);
+
+   private:
+    struct vtable_t {
+        void (*dtor)(void*) noexcept;
+        void (*copy)(void const*, void*);
+        void (*move)(void*, void*) noexcept;
+        void (*to_proto_value)(void const*, google::protobuf::Value*);
+        int (*kind)() noexcept;
+        bool (*equal_to)(void const*, void const*, const vtable_t&);
+    };
+
+    template <typename T>
+    struct model_t {
+        model_t(T t) : data(std::move(t)) {}
+
+        static void dtor(void* self) noexcept {
+            static_cast<model_t*>(self)->~model_t();
+        }
+
+        static void copy(void const* self, void* dest) {
+            new (dest) model_t(*static_cast<model_t const*>(self));
```

Comment:
I'm having vague thoughts about std::launder here due to seeing placement new, but I need to remind myself about what that does and how it works.

Reply:
I was wondering about that myself, especially since std::launder didn't exist pre-c++17, though people were implementing variant and optional types all the while. This discussion from the boost mailing list suggests we're probably fine:
https://groups.google.com/g/boost-developers-archive/c/-ivUay0RxQw